### PR TITLE
python3Packages.pyarrow: enable dataset and flight modules

### PR DESCRIPTION
**Pull request by cpcloud**
- python3Packages.pyarrow: enable dataset module
- python3Packages.pyarrow: add cpcloud as maintainer
- python3Packages.pyarrow: enable flight module

###### Motivation for this change

Enable the `_dataset` and `_flight` modules in `pyarrow`

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by veprbl** [comment]
@GrahamcOfBorg build python38Packages.pyarrow python39Packages.pyarrow

File: pkgs/development/python-modules/pyarrow/default.nix [hunk 1]
```diff
@@ -19,6 +19,7 @@ buildPythonPackage rec {
   PYARROW_BUILD_TYPE = "release";
 
   PYARROW_WITH_DATASET = true;
+  PYARROW_WITH_FLIGHT = true;
```
Comment (veprbl) [inline pkgs/development/python-modules/pyarrow/default.nix:22]: We need to inherit this value from the enableFlight value passed to arrow-cpp

Comment (cpcloud) [reply pkgs/development/python-modules/pyarrow/default.nix:22]: Is that value accessible? I'm not aware of a way to access it

Comment (veprbl) [reply pkgs/development/python-modules/pyarrow/default.nix:22]: You will need to add something like `passthru = { inherit enableFlight; }` to the expression for arrow-cpp

**Comment by cpcloud** [comment]
@veprbl Thanks, done.

**Review by veprbl — APPROVED** [review]

**Comment by veprbl** [comment]
@cpcloud  Could you please squash the last three comments.

**Pull request merged**